File: backend/ria_ml/combinedLockboxes_process.py
import numpy as np
import matplotlib.pyplot as plt

def combinedLockboxes_process(combinedLockboxes, client):
    wts = np.array(combinedLockboxes['componentWeights'], dtype=float)
    wts = np.maximum(wts, 0)
    wts = wts / np.sum(wts)

    boxprops = np.array(combinedLockboxes['componentLockboxes'][0]['proportions'])
    combprops = wts[0] * boxprops
    for i in range(1, len(combinedLockboxes['componentLockboxes'])):
        boxprops = np.array(combinedLockboxes['componentLockboxes'][i]['proportions'])
        combprops = combprops + wts[i] * boxprops
    combinedLockboxes['proportions'] = combprops

    xs = combprops
    nyrs = xs.shape[1]
    if combinedLockboxes['showCombinedProportions'].lower() == 'y':
        fig, ax = plt.subplots()
        x = np.arange(1, nyrs + 1)
        ax.bar(x, xs[0, :], label='TIPS')
        ax.bar(x, xs[1, :], bottom=xs[0, :], label='Market')
        ax.grid(True)
        ax.set_xlabel('Lockbox Maturity Year')
        ax.set_ylabel('Amount Invested at Inception')
        ax.legend(['TIPS', 'Market'])
        ax.set_xlim([0, nyrs + 1])
        ax.set_title(f"Lockbox Proportions for {combinedLockboxes['title']}", color='b')
        plt.show()

    return combinedLockboxes

File: backend/ria_ml/combinedLockboxes_process.m
function combinedLockboxes = combinedLockboxes_process(combinedLockboxes,client);
  
  % combines componentLockboxes in combinedLockboxes to create a new lockbox
    n = length(combinedLockboxes.componentLockboxes);
    wts = combinedLockboxes.componentWeights;
    wts =  max(wts, 0);
    wts = wts / sum(wts);
    
    boxprops = combinedLockboxes.componentLockboxes{1}.proportions;
    combprops  = wts(1) * boxprops;
    for i = 2:length(combinedLockboxes.componentLockboxes)
        boxprops = combinedLockboxes.componentLockboxes{i}.proportions;
        combprops  = combprops + ( wts(i) * boxprops);
    end; 
    combinedLockboxes.proportions =  combprops; 
    
  % plot contents if requested
     xs =  combinedLockboxes.proportions;
     nyrs = size(xs,2);
     if lower(combinedLockboxes.showCombinedProportions) == 'y'
        fig = figure;
        x = 1:1:size(xs,2);
        bar(x,xs','stacked'); grid;
        set(gca,'FontSize',30);
        ss = client.figurePosition;
        set(gcf,'Position',ss);
        set(gcf,'Color',[1 1 1]);
        xlabel('Lockbox Maturity Year ','fontsize',30);
        ylabel('Amount Invested at Inception   ','fontsize',30);
        legend('TIPS ','Market ');
        ax = axis; ax(1) = 0; ax(2) = nyrs+1; ax(3) = 0; ax(4) = 1; axis(ax);
        t = ['Lockbox Proportions for ' combinedLockboxes.title];
        title(t,'Fontsize',40,'Color','b');
        beep; pause;
     end; %if lower(combinedLockboxes.showContents) = 'y'    
     
end % function 
